### PR TITLE
SYS-4002: Fix primary validator for summary operations

### PR DESCRIPTION
**Pull request by micaelffrancoAV**
## Proposed changes

- PR to change avn primary validator to be calculated based on the block number.

## Type of change/Merge

🚨What type of change is this PR? </br>
_Put an `x` in the boxes that apply_

- [ ] Release 
  - [ ] Increase versions 
  - [ ] Baseline tests passed  
  - Release type:
    - [ ] Major release 
    - [ ] Minor release 
    - [ ] Patch release 

## Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR._

- [ ] You describe the purpose of the PR, e.g.:
  - What does it do?
  - Highlight what important points reviewers should know about;
  - Indicates if there is something left for follow-up PRs.
- [ ] Documentation updated
- [ ] Business logic tested successfully
- [ ] Verify First, Write Last: In Substrate development, it is important that you always ensure preconditions are met and return errors at the beginning. After these checks have completed, then you may begin the function's computation.

### CODE REVIEW

File: pallets/avn/src/lib.rs [hunk 1]
```diff
@@ -181,20 +175,14 @@ pub mod pallet {
 
     #[pallet::storage]
     #[pallet::getter(fn get_primary_collator)]
-    pub type PrimaryCollator<T: Config> = StorageValue<_, PrimaryCollatorData, ValueQuery>;
+    pub type PrimaryCollator<T: Config> = StorageValue<_, u8, ValueQuery>;
```
Comment (nahuseyoum) [inline pallets/avn/src/lib.rs:178]: Can we rename this to make it more explicit? Something like `primaryCollatorIndexForEth`? I also think we should use a large type for the value. u8 give 255 only and its possible we will end up sending more than that. Perhaps a `u32`?

**Review by nahuseyoum — APPROVED** [review]

File: pallets/ethereum-events/src/lib.rs [hunk 2]
```diff
@@ -792,15 +791,15 @@ pub mod pallet {
             let (this_validator, finalised_block) = setup_result.expect("We have a validator");
 
             // Only primary validators can check and process events
-            let is_primary =
-                AVN::<T>::is_primary(OperationType::Ethereum, &this_validator.account_id);
-            if is_primary.is_err() {
+            let is_primary_validator_for_sending =
```
Comment (nahuseyoum) [inline pallets/ethereum-events/src/lib.rs:794]:
```suggestion
            let is_primary =
```

File: pallets/ethereum-events/src/lib.rs [hunk 3]
```diff
@@ -836,7 +835,7 @@ pub mod pallet {
                     return InvalidTransaction::Custom(ERROR_CODE_INVALID_EVENT_DATA).into()
                 }
 
-                if AVN::<T>::is_primary(OperationType::Ethereum, &result.checked_by)
+                if AVN::<T>::is_primary_validator_for_sending(&result.checked_by)
```
Comment (nahuseyoum) [inline pallets/ethereum-events/src/lib.rs:838]:
```suggestion
                if AVN::<T>::is_primary_for_block(&result.checked_by)
```

File: pallets/ethereum-events/src/mock.rs [hunk 4]
```diff
@@ -412,11 +412,10 @@ impl EthereumEvents {
         return AVN::active_validators()
     }
 
-    pub fn is_primary(
-        op_type: OperationType,
+    pub fn is_primary_validator_for_sending(
         validator: &AccountId,
     ) -> Result<bool, avn_error<TestRuntime>> {
-        return AVN::is_primary(op_type, validator)
+        return AVN::is_primary_validator_for_sending(validator)
```
Comment (nahuseyoum) [inline pallets/ethereum-events/src/mock.rs:418]:
```suggestion
        return AVN::is_primary_for_block(validator)
```

File: pallets/ethereum-events/src/tests/tests.rs [hunk 5]
```diff
@@ -190,7 +190,7 @@ fn test_is_primary_blocknumber_3() {
     ext.execute_with(|| {
         let block_number = 3;
         let expected_primary = account_id_1();
-        let result = EthereumEvents::is_primary(OperationType::Ethereum, &expected_primary);
+        let result = EthereumEvents::is_primary_validator_for_sending(&expected_primary);
```
Comment (nahuseyoum) [inline pallets/ethereum-events/src/tests/tests.rs:193]:
```suggestion
        let result = EthereumEvents::is_primary_for_block(&expected_primary);
```

File: pallets/ethereum-events/src/tests/tests.rs [hunk 6]
```diff
@@ -312,7 +312,7 @@ fn is_primary_fails_with_no_validators() {
     let mut ext = ExtBuilder::build_default().as_externality();
     ext.execute_with(|| {
         let block_number = 1;
-        let result = EthereumEvents::is_primary(OperationType::Ethereum, &account_id_1());
+        let result = EthereumEvents::is_primary_validator_for_sending(&account_id_1());
```
Comment (nahuseyoum) [inline pallets/ethereum-events/src/tests/tests.rs:315]:
```suggestion
        let result = EthereumEvents::is_primary_for_block(&account_id_1());
```

File: pallets/ethereum-events/src/tests/tests.rs [hunk 7]
```diff
@@ -630,7 +630,7 @@ mod signature_in {
             UintAuthorityId::set_all_keys(vec![1, 2, 3]);
 
             let val_length = EthereumEvents::validators().len();
-            let index_of_primary_validator = AVN::<TestRuntime>::get_primary_collator().ethereum;
+            let index_of_primary_validator = AVN::<TestRuntime>::get_primary_collator();
```
Comment (nahuseyoum) [inline pallets/ethereum-events/src/tests/tests.rs:633]: Since we are not sending here, shouldn't we be using the block number way?

File: pallets/summary/src/challenge.rs [hunk 8]
```diff
@@ -135,8 +134,9 @@ fn can_challenge<T: Config>(
         return false
     }
 
-    let is_chosen_validator = AVN::<T>::is_primary(OperationType::Avn, &this_validator.account_id)
-        .unwrap_or_else(|_| false);
+    let is_chosen_validator =
+        AVN::<T>::is_primary_validator_on_block(ocw_block_number, &this_validator.account_id)
```
Comment (nahuseyoum) [inline pallets/summary/src/challenge.rs:138]: :+1:

File: pallets/summary/src/lib.rs [hunk 9]
```diff
@@ -751,7 +751,7 @@ pub mod pallet {
                 safe_add_block_numbers::<BlockNumberFor<T>>(Self::current_slot(), 1u32.into())
                     .map_err(|_| Error::<T>::Overflow)?;
 
-            let new_validator_account_id = AVN::<T>::advance_primary_validator(OperationType::Avn)?;
+            let new_validator_account_id = AVN::<T>::advance_primary_validator_for_sending()?;
```
Comment (nahuseyoum) [inline pallets/summary/src/lib.rs:754]: I don't think we should be doing that here. The index is only incremented when sending I think.

File: pallets/summary/src/tests/tests_challenge.rs [hunk 10]
```diff
@@ -152,19 +159,20 @@ mod challenge_slot_if_required {
                 assert!(pool_state.read().transactions.is_empty());
 
                 System::set_block_number(context.block_after_grace_period);
+                // let challenge = get_valid_challenge(&context);
```
Comment (nahuseyoum) [inline pallets/summary/src/tests/tests_challenge.rs:162]:
```suggestion               
```

File: pallets/summary/src/tests/tests_vote.rs [hunk 11]
```diff
@@ -897,7 +897,7 @@ mod end_voting_period {
                 Summary::set_previous_summary_slot(5);
 
                 let primary_validator_id =
-                    AVN::<TestRuntime>::advance_primary_validator(OperationType::Avn)
+                    AVN::<TestRuntime>::advance_primary_validator_for_sending()
```
Comment (nahuseyoum) [inline pallets/summary/src/tests/tests_vote.rs:900]: do we need to advance here? I think only sending should advance.

File: pallets/avn/src/tests/tests.rs [hunk 12]
```diff
@@ -131,13 +58,13 @@ mod calling_is_primary_validator_for_ethereum {
         let mut ext = ExtBuilder::build_default().with_validators().as_externality();
         ext.execute_with(|| {
             let expected_primary = 1;
-            let eth_index_before: u8 = get_index_based_on_operation_type(&OperationType::Ethereum);
+            let eth_index_before: u8 = AVN::get_primary_collator();
```
Comment (nahuseyoum) [inline pallets/avn/src/tests/tests.rs:61]: where is this defined?

File: pallets/eth-bridge/src/tx.rs [hunk 13]
```diff
@@ -143,7 +141,7 @@ fn generate_msg_hash<T: pallet::Config>(
 }
 
 fn assign_sender<T: Config>() -> Result<T::AccountId, Error<T>> {
-    match AVN::<T>::advance_primary_validator(OperationType::Ethereum) {
+    match AVN::<T>::advance_primary_validator_for_sending() {
```
Comment (nahuseyoum) [inline pallets/eth-bridge/src/tx.rs:144]: :+1:

File: pallets/avn/src/tests/tests.rs (hunk 12 above)
Comment (micaelffrancoAV) [reply pallets/avn/src/tests/tests.rs:61]: Its the getter for the primary collator storage index.

**Review by ivan-cholakov — APPROVED** [review]

File: Cargo.toml [hunk 14]
```diff
@@ -12,7 +12,7 @@ lto = "fat"
 codegen-units = 1
 
 [workspace.package]
-version = "5.2.3"
+version = "5.2.4"
```
Comment (nahuseyoum) [inline Cargo.toml:15]: and this

File: runtime/avn/src/lib.rs [hunk 15]
```diff
@@ -174,7 +174,7 @@ pub const VERSION: RuntimeVersion = RuntimeVersion {
     spec_name: create_runtime_str!("avn-parachain"),
     impl_name: create_runtime_str!("avn-parachain"),
     authoring_version: 1,
-    spec_version: 62,
+    spec_version: 63,
```
Comment (nahuseyoum) [inline runtime/avn/src/lib.rs:177]: Can we revert this change please?

File: runtime/test/src/lib.rs [hunk 16]
```diff
@@ -169,7 +169,7 @@ pub const VERSION: RuntimeVersion = RuntimeVersion {
     spec_name: create_runtime_str!("avn-test-parachain"),
     impl_name: create_runtime_str!("avn-test-parachain"),
     authoring_version: 1,
-    spec_version: 62,
+    spec_version: 63,
```
Comment (nahuseyoum) [inline runtime/test/src/lib.rs:172]: and this too

**Review by nahuseyoum — APPROVED** [review]
